File: source/go/producer_consumer.go
package main

import (
	"fmt"
	"strings"
)

func producer(ch chan<- int) {
	for i := 1; i <= 10; i++ {
		ch <- i
	}
	close(ch)
}

func consumer(ch <-chan int) []int {
	var results []int
	for val := range ch {
		results = append(results, val)
	}
	return results
}

func main() {
	ch := make(chan int, 10)
	go producer(ch)
	results := consumer(ch)

	parts := make([]string, len(results))
	for i, v := range results {
		parts[i] = fmt.Sprintf("%d", v)
	}
	fmt.Printf("received: %s\n", strings.Join(parts, " "))
}

File: source/python/producer_consumer.py
import threading
import queue

ch = queue.Queue(maxsize=10)

def producer():
    for i in range(1, 11):
        ch.put(i)
    ch.put(None)  # sentinel

def consumer():
    results = []
    while True:
        val = ch.get()
        if val is None:
            break
        results.append(val)
    return results

results_holder = []

def consumer_thread():
    results_holder.extend(consumer())

p = threading.Thread(target=producer)
c = threading.Thread(target=consumer_thread)

p.start()
c.start()
p.join()
c.join()

print(f"received: {' '.join(str(x) for x in results_holder)}")
